### PR TITLE
feat(lambda-event-sources): batching window for sqs event source

### PR DESCRIPTION
**Pull request by SeekerWing**
fixes https://github.com/aws/aws-cdk/issues/11722

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 1]
```diff
@@ -29,14 +39,20 @@ export class SqsEventSource implements lambda.IEventSource {
...
   constructor(readonly queue: sqs.IQueue, private readonly props: SqsEventSourceProps = { }) {
-    if (this.props.batchSize !== undefined && (this.props.batchSize < 1 || this.props.batchSize > 10)) {
-      throw new Error(`Maximum batch size must be between 1 and 10 inclusive (given ${this.props.batchSize})`);
+    if (this.props.batchSize !== undefined) {
+      if (this.props.maxBatchingWindow !== undefined && (this.props.batchSize < 1 || this.props.batchSize > 10000)) {
+        throw new Error(`Maximum batch size must be between 1 and 10000 inclusive (given ${this.props.batchSize})`);
+      }
+      if (this.props.maxBatchingWindow === undefined && (this.props.batchSize < 1 || this.props.batchSize > 10)) {
+        throw new Error(`Maximum batch size must be between 1 and 10 inclusive (given ${this.props.batchSize})`);
+      }
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:48]: Just to clarify what you're doing here - max batch size should not be between 1 and 10 if max batching window is specified. Is that correct and intended?

Comment (nija-at) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:48]: If so, maybe update the message to 

```suggestion
      if (this.props.maxBatchingWindow === undefined && (this.props.batchSize < 1 || this.props.batchSize > 10)) {
        throw new Error(`Maximum batch size must be between 1 and 10 inclusive (given ${this.props.batchSize}) when batching window is specified.`);
      }
```

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 2]
```diff
@@ -14,6 +14,16 @@ export interface SqsEventSourceProps {
    */
   readonly batchSize?: number;
 
+  /**
+   * TThe maximum amount of time to gather records before invoking the function.
+   * Maximum of Duration.minutes(5)
+   *
+   * Valid Range: Minimum value of 0 minutes. Maximum value of 5 minutes.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:21]: Add validation for this in the constructor.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 3]
```diff
@@ -29,14 +39,20 @@ export class SqsEventSource implements lambda.IEventSource {
   private _eventSourceMappingId?: string = undefined;
 
   constructor(readonly queue: sqs.IQueue, private readonly props: SqsEventSourceProps = { }) {
-    if (this.props.batchSize !== undefined && (this.props.batchSize < 1 || this.props.batchSize > 10)) {
-      throw new Error(`Maximum batch size must be between 1 and 10 inclusive (given ${this.props.batchSize})`);
+    if (this.props.batchSize !== undefined) {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:42]: Looks like we've missed to check for tokens here. Could you add that, and a corresponding unit test?

Comment (nija-at) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:42]:
```suggestion
    if (this.props.batchSize !== undefined && !Token.isUnresolved(this.props.batchSize)) {
```

File: packages/@aws-cdk/aws-lambda-event-sources/test/integ.sqs.ts [hunk 4]
```diff
@@ -12,6 +12,7 @@ class SqsEventSourceTest extends cdk.Stack {
 
     fn.addEventSource(new SqsEventSource(queue, {
       batchSize: 5,
+      maxBatchingWindow: cdk.Duration.minutes(3),
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/test/integ.sqs.ts:15]: No need to update the integ test for this change.

**Review by nija-at — DISMISSED** [review]
Thanks for submitting a PR for this change. I've added my comments below.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 5]
```diff
@@ -14,6 +14,16 @@ export interface SqsEventSourceProps {
    */
   readonly batchSize?: number;
 
+  /**
+   * TThe maximum amount of time to gather records before invoking the function.
```
Comment (jogold) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:18]:
```suggestion
   * The maximum amount of time to gather records before invoking the function.
```

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 6]
```diff
@@ -29,14 +39,20 @@ export class SqsEventSource implements lambda.IEventSource {
   private _eventSourceMappingId?: string = undefined;
 
   constructor(readonly queue: sqs.IQueue, private readonly props: SqsEventSourceProps = { }) {
```
Comment (jogold) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:41]: Ideally you should also throw here `if (this.queue.fifo && this.props.maxBatchingWindow)`

Comment (SeekerWing) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:41]: thanks for the suggestion @jogold, nice catch

adding [doc reference](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html) for posterity. 

> Batch window – Specify the maximum amount of time to gather records before invoking the function, in seconds. Only applicable to standard queues.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 7]
```diff
@@ -14,6 +14,16 @@ export interface SqsEventSourceProps {
    */
   readonly batchSize?: number;
 
+  /**
+   * The maximum amount of time to gather records before invoking the function.
+   * Maximum of Duration.minutes(5)
+   *
+   * Valid Range: Minimum value of 0 minutes. Maximum value of 5 minutes.
+   *
+   * @default Duration.seconds(0)
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:23]: Better to say something like this -

```suggestion
   * @default - no batching window. The lambda function will be invoked immediately with the records that are available.
```

File: packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts [hunk 8]
```diff
@@ -29,14 +39,30 @@ export class SqsEventSource implements lambda.IEventSource {
...
   constructor(readonly queue: sqs.IQueue, private readonly props: SqsEventSourceProps = { }) {
-    if (this.props.batchSize !== undefined && (this.props.batchSize < 1 || this.props.batchSize > 10)) {
-      throw new Error(`Maximum batch size must be between 1 and 10 inclusive (given ${this.props.batchSize})`);
+    const maxBatchingWindow = this.props.maxBatchingWindow;
+    if (maxBatchingWindow !== undefined) {
+      if (queue.fifo) {
+        throw new Error('Batching window is not supported for FIFO queues');
+      }
+      if (maxBatchingWindow.toSeconds() < 0 || maxBatchingWindow.toSeconds() > 300) {
+        throw new Error(`Maximum batching window must be between 0 and 300 seconds (given ${maxBatchingWindow.toSeconds()})`);
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/sqs.ts:48]: AFAIK, it's not possible to construct a `Duration` that is less than 0 seconds.

```suggestion
      if (maxBatchingWindow.toSeconds() > 300) {
        throw new Error(`Maximum batching window must be 300 seconds or less (given ${maxBatchingWindow.toSeconds()})`);
```

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts [hunk 9]
```diff
@@ -85,6 +85,38 @@ export = {
     test.done();
   },
 
+  'unresolved batch size'(test: Test) {
+    // GIVEN
+    const stack = new cdk.Stack();
+    const fn = new TestFunction(stack, 'Fn');
+    const q = new sqs.Queue(stack, 'Q');
+    let batchSize : number;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/test/test.sqs.ts:93]:
```suggestion
    const batchSize: number = 500;
```

**Review by nija-at — CHANGES_REQUESTED** [review]
Thanks for fixing this.

This looks ready. Just a couple of comments below to improve error messaging and documentation.

**Comment by cbenning** [comment]
What is the status on this making it into CDK? my team is waiting on this new feature.

**Comment by nija-at** [comment]
Looks like this PR has not been worked on for several months.

@cbenning - if you're interested, you could pick this up and get it past the finish line.

**Comment by SeekerWing** [comment]
closing this PR in favor of https://github.com/aws/aws-cdk/pull/13406

**Pull request closed without merging**